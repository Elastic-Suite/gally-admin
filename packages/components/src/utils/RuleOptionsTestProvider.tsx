import React, { ReactNode } from 'react'
import {
  IRuleEngineOperators,
  RuleAttributeType,
  getIri,
} from '@elastic-suite/gally-admin-shared'

import ruleEngineOperator from '../../public/mocks/rule_engine_operators.json'
import sourceFields from '../../public/mocks/source_fields.json'
import sourceFieldLabels from '../../public/mocks/source_field_labels.json'

import RuleOptionsProvider from '../components/stateful-providers/RuleOptionsProvider/RuleOptionsProvider'

interface IProps {
  children: ReactNode
}

const sourceFieldLabelsMap = new Map(
  sourceFieldLabels['hydra:member'].map((label) => [
    label.sourceField,
    label.label,
  ])
)

const fields = sourceFields['hydra:member'].map((field) => {
  const label = sourceFieldLabelsMap.get(getIri('source_fields', field.id))
  return {
    id: field.id,
    code: field.code,
    label: (label || field.defaultLabel) ?? '',
    type: field.type as RuleAttributeType,
  }
})

function RuleOptionsTestProvider(props: IProps): JSX.Element {
  const { children } = props
  return (
    <RuleOptionsProvider
      catalogId={-1}
      localizedCatalogId={-1}
      fields={fields}
      ruleOperators={ruleEngineOperator as IRuleEngineOperators}
    >
      {children}
    </RuleOptionsProvider>
  )
}

export default RuleOptionsTestProvider
