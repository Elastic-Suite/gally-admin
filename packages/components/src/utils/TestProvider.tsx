import React, { ReactNode } from 'react'

import catalogs from '../../public/mocks/catalog.json'
import metadata from '../../public/mocks/metadata.json'

import { catalogContext, metadataContext } from '../contexts'
import { useCatalogs } from '../hooks'

import OptionsProvider from '../components/stateful-providers/OptionsProvider/OptionsProvider'

interface IProps {
  children: ReactNode
}

function TestProvider(props: IProps): JSX.Element {
  const { children } = props
  const contextValueCatalog = useCatalogs(catalogs['hydra:member'])
  const contextValueMetadata = metadata['hydra:member']

  return (
    <OptionsProvider>
      <catalogContext.Provider value={contextValueCatalog}>
        <metadataContext.Provider value={contextValueMetadata}>
          {children}
        </metadataContext.Provider>
      </catalogContext.Provider>
    </OptionsProvider>
  )
}

export default TestProvider
