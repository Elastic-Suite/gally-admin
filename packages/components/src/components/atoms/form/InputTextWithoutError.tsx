import React, {
  ChangeEvent,
  ForwardedRef,
  HTMLAttributes,
  ReactNode,
  Ref,
  SyntheticEvent,
  forwardRef,
} from 'react'
import { FormHelperText, InputLabel } from '@mui/material'

import { getFormValue } from '../../../services'

import IonIcon from '../IonIcon/IonIcon'

import InfoTooltip from './InfoTooltip'
import {
  IUnstyledInputTextProps,
  InputTextStyled,
  StyledFormControl,
  Suffix,
  Wrapper,
} from './InputText.styled'

import { TestId, generateTestId } from '../../../utils/testIds'

export interface IInputTextProps
  extends Omit<
      IUnstyledInputTextProps,
      'margin' | 'onChange' | 'onKeyUp' | 'onKeyDown' | 'onBlur' | 'onFocus'
    >,
    Pick<
      HTMLAttributes<HTMLInputElement>,
      'onKeyUp' | 'onKeyDown' | 'onBlur' | 'onFocus'
    > {
  error?: boolean
  fullWidth?: boolean
  infoTooltip?: string
  inputRef?: Ref<HTMLInputElement>
  label?: ReactNode
  margin?: 'none' | 'dense' | 'normal'
  helperText?: ReactNode
  helperIcon?: string
  onChange?: (value: string | number, event: SyntheticEvent) => void
  suffix?: ReactNode
  requiredLabel?: boolean
  componentId?: string
}

function InputTextWithoutError(
  props: IInputTextProps,
  ref?: ForwardedRef<HTMLDivElement>
): JSX.Element {
  const {
    error,
    fullWidth,
    helperText,
    helperIcon,
    id,
    infoTooltip,
    label,
    margin,
    onChange,
    required,
    suffix,
    value,
    requiredLabel,
    componentId,
    inputProps,
    ...InputProps
  } = props

  function handleChange(
    event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ): void {
    const { value } = event.target
    if (onChange) {
      onChange(getFormValue(value, props), event)
    }
  }

  return (
    <StyledFormControl
      error={error}
      fullWidth={fullWidth}
      margin={margin}
      variant="standard"
    >
      {Boolean(label || infoTooltip) && (
        <InputLabel shrink htmlFor={id} required={requiredLabel || required}>
          {label}
          {infoTooltip ? <InfoTooltip title={infoTooltip} /> : null}
        </InputLabel>
      )}
      <Wrapper className="InputText__Wrapper">
        <InputTextStyled
          id={id}
          onChange={handleChange}
          required={required}
          ref={ref}
          value={(value ?? '').toString()}
          inputProps={{
            'data-testid': generateTestId(TestId.INPUT_TEXT, componentId),
            ...inputProps,
          }}
          {...InputProps}
        />
        {Boolean(suffix) && <Suffix>{suffix}</Suffix>}
      </Wrapper>
      {Boolean(helperText) && (
        <FormHelperText
          data-testid={generateTestId(TestId.HELPER_TEXT, componentId)}
        >
          {Boolean(helperIcon) && (
            <IonIcon
              name={helperIcon}
              style={{ fontSize: 18, marginRight: 2 }}
            />
          )}
          {helperText}
        </FormHelperText>
      )}
    </StyledFormControl>
  )
}

export default forwardRef<HTMLDivElement, IInputTextProps>(
  InputTextWithoutError
)
