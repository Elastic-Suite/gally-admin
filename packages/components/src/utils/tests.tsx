import React, { ReactNode } from 'react'
import { RenderOptions, render, renderHook } from '@testing-library/react'
import { PreloadedState } from '@reduxjs/toolkit'
import { Bundle, api } from '@elastic-suite/gally-admin-shared'

import { AppStore, RootState, setupStore } from '../store'

import AppProvider from '../components/stateful-providers/AppProvider/AppProvider'

import TestProvider from './TestProvider'

export interface IExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export interface IRenderOutput {
  store: AppStore
}

const defaultState = {
  data: {
    api,
    bundles: [Bundle.VIRTUAL_CATEGORY],
    configurations: {
      'base_url/media': 'https://localhost/media/catalog/product/',
    },
    metadata: [
      {
        '@id': '/metadata/3',
        '@type': 'Metadata',
        id: 3,
        entity: 'product',
        sourceFields: ['/source_fields/217', '/source_fields/253'],
      },
      {
        '@id': '/metadata/4',
        '@type': 'Metadata',
        id: 4,
        entity: 'category',
        sourceFields: ['/source_fields/254', '/source_fields/286'],
      },
    ],
  },
  user: {
    requestedPath: '',
  },
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = defaultState,
    store,
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  store = store ?? setupStore(preloadedState)
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <AppProvider store={store}>
        <TestProvider>{children}</TestProvider>
      </AppProvider>
    )
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderHookWithProviders<R>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  hook: (props: any) => R,
  {
    preloadedState = defaultState,
    store,
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  store = store ?? setupStore(preloadedState)
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <AppProvider store={store}>
        <TestProvider>{children}</TestProvider>
      </AppProvider>
    )
  }
  return { store, ...renderHook(hook, { wrapper: Wrapper, ...renderOptions }) }
}
