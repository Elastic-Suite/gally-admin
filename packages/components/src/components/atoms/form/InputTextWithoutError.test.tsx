import React from 'react'

import { renderWithProviders } from '../../../utils/tests'

import InputTextWithoutError from './InputTextWithoutError'

describe('InputTextWithoutError match snapshot', () => {
  it('BadgeDisabledFalse', () => {
    const { container } = renderWithProviders(
      <InputTextWithoutError
        id="input-text"
        label="Label"
        placeholder="Name"
        required={false}
        disabled={false}
        helperText=""
        helperIcon=""
        color="primary"
        endAdornment={null}
        value="hello world"
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('BadgeDisabledFalseRequired', () => {
    const { container } = renderWithProviders(
      <InputTextWithoutError
        id="input-text"
        label="Label"
        placeholder="Name"
        required
        disabled
        helperText=""
        helperIcon=""
        color="primary"
        endAdornment={null}
        value="hello world"
      />
    )
    expect(container).toMatchSnapshot()
  })
})
