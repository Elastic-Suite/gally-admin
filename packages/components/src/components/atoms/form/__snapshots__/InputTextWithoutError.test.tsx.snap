// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputTextWithoutError match snapshot BadgeDisabledFalse 1`] = `
<div>
  <div
    class="MuiFormControl-root css-hh3vi8-MuiFormControl-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled css-rnc4nv-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for="input-text"
    >
      Label
    </label>
    <div
      class="InputText__Wrapper css-17fp4jk"
    >
      <div
        class="MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1jf161a-MuiInputBase-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input css-ngazp5-MuiInputBase-input"
          data-testid="inputText"
          id="input-text"
          placeholder="Name"
          type="text"
          value="hello world"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`InputTextWithoutError match snapshot BadgeDisabledFalseRequired 1`] = `
<div>
  <div
    class="MuiFormControl-root css-hh3vi8-MuiFormControl-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required css-rnc4nv-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for="input-text"
    >
      Label
      <span
        aria-hidden="true"
        class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1ufxisk-MuiFormLabel-asterisk"
      >
        â€‰
        *
      </span>
    </label>
    <div
      class="InputText__Wrapper css-17fp4jk"
    >
      <div
        class="MuiInputBase-root MuiInputBase-colorPrimary Mui-disabled MuiInputBase-formControl css-1jf161a-MuiInputBase-root"
        required=""
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input Mui-disabled css-ngazp5-MuiInputBase-input"
          data-testid="inputText"
          disabled=""
          id="input-text"
          placeholder="Name"
          required=""
          type="text"
          value="hello world"
        />
      </div>
    </div>
  </div>
</div>
`;
